core/vm/validate.go

```diff
@@ -0,0 +1,343 @@
+			}
+			if ct := container.ContainerSections[arg]; len(ct.Data) != ct.DataSize {
+				return nil, fmt.Errorf("%w: container %d, have %d, claimed %d, pos %d", ErrEOFCreateWithTruncatedSection, arg, len(ct.Data), ct.DataSize, i)
+			}
+			if _, ok := visitedSubcontainers[arg]; ok {
+				return nil, fmt.Errorf("section already referenced, arg :%d", arg)
+			}
+			// We need to store per subcontainer how it was referenced
+			if v, ok := visitedSubcontainers[arg]; ok && v != RefByEOFCreate {
+				return nil, fmt.Errorf("section already referenced, arg :%d", arg)
+			}
+			visitedSubcontainers[arg] = RefByEOFCreate
+		case op == STOP || op == RETURN:
+			if isInitCode {
+				return nil, ErrStopInInitCode
+			}
+			if hasReturnContract {
+				return nil, ErrStopAndReturnContract
+			}
+			hasStop = true
+		}
+		i += size + 1
+	}
+	// Code sections may not "fall through" and require proper termination.
+	// Therefore, the last instruction must be considered terminal or RJUMP.
+	if !jt[op].terminal && op != RJUMP {
+		return nil, fmt.Errorf("%w: end with %s, pos %d", ErrInvalidCodeTermination, op, i)
+	}
+	if paths, err := validateControlFlow2(code, section, container.Types, jt); err != nil {
+		return nil, err
+	} else if paths != count {
+		// TODO(matt): return actual position of unreachable code
+		return nil, ErrUnreachableCode
+	}
+	return &ValidationResult{
+		VisitedCode:          visitedCode,
+		VisitedSubContainers: visitedSubcontainers,
+		IsInitCode:           hasReturnContract,
+		IsRuntime:            hasStop,
+	}, nil
+}
+
+// checkDest parses a relative offset at code[0:2] and checks if it is a valid jump destination.
+func checkDest(code []byte, analysis *bitvec, imm, from, length int) error {
+	if len(code) < imm+2 {
+		return io.ErrUnexpectedEOF
+	}
+	if analysis != nil && *analysis == nil {
+		*analysis = eofCodeBitmap(code)
+	}
+	offset := parseInt16(code[imm:])
+	dest := from + offset
+	if dest < 0 || dest >= length {
+		return fmt.Errorf("%w: out-of-bounds offset: offset %d, dest %d, pos %d", ErrInvalidJumpDest, offset, dest, imm)
+	}
+	if !analysis.codeSegment(uint64(dest)) {
+		return fmt.Errorf("%w: offset into immediate: offset %d, dest %d, pos %d", ErrInvalidJumpDest, offset, dest, imm)
+	}
+	return nil
+}
+
+// validateControlFlow iterates through all possible branches the provided code
+// value and determines if it is valid per EOF v1.
+func validateControlFlow(code []byte, section int, metadata []*FunctionMetadata, jt *JumpTable) (int, error) {
+	type item struct {
+		pos       int
+		height    int
+		backwards bool
+	}
+	var (
+		heights        = make(map[int]int)
+		worklist       = []item{{0, int(metadata[section].Input), false}}
+		maxStackHeight = int(metadata[section].Input)
+	)
+	for 0 < len(worklist) {
+		var (
+			idx       = len(worklist) - 1
+			pos       = worklist[idx].pos
+			height    = worklist[idx].height
+			backwards = worklist[idx].backwards
+		)
+		worklist = worklist[:idx]
+	outer:
```

Comment:
Whenever we have these loops and labels and `break outer`, it's IMO usually better to branch off a method. So instead of
```
for ....
  outer: 
    for .. 
      switch 
         break outer
         ....
      break outer
 bazonk()
```
we do 
```
for ....
  inner()
 bazonk()


func inner(){
    for .. 
      switch 
         return //break outer
         ....
      return // break outer
}
```

Reply:
ah, this is old unused code. Please just nuke it, if we're not using it.